### PR TITLE
pref center work in progress

### LINKED ISSUE #9994: A couple of Mautic specific building blocks such as segment list & save preferences are missing TPROD-208
As the title mentions, A couple of Mautic specific building blocks such as segment list & save preferences are missing.
I am wondering if this is somewhere on the roadmap, a patch that I can help test or if this is still in the "needs help" phase? Pointers are very welcome, so maybe I/my team can help.

The existing mautic builder
image

The new grapesjs builder
image

We're currently using the existing builder and this gives us the following HTML code for a preference center https://github.com/mautic/mautic/tree/37ed5ee87900522c888b09ed3857ad0452fb33df/app/bundles/CoreBundle/Views/Slots
See https://github.com/mautic/mautic/blob/37ed5ee87900522c888b09ed3857ad0452fb33df/app/bundles/CoreBundle/Views/Slots/saveprefsbutton.html.php & https://github.com/mautic/mautic/blob/37ed5ee87900522c888b09ed3857ad0452fb33df/app/bundles/CoreBundle/Views/Slots/segmentlist.html.php

Am I missing something as I don't see those slots in grapesjs? I also can't see them in https://github.com/mautic/grapesjs-preset-mautic 

I think, in order to mark this as stable, the slots in core mautic should at least be supported no?

### PR DESCRIPTION
**Pull request by disha1611**
⚠️ 🚧  This is a WIP PR 🚧  ⚠️ 

This PR will add the capability to configure the preference centre options in landing pages with the GrapesJS editor.

It fixes https://github.com/mautic/mautic/issues/9994.

### CODE REVIEW

File: dist/blocks.js [hunk 1]
```diff
@@ -7,10 +10,68 @@ export default ((editor, opts = {}) => {
   if ('grapesjsmjml' in cfg.pluginsOpts) {// Dynamic Content MJML block
   } else if ('grapesjsnewsletter' in cfg.pluginsOpts) {
     const dcb = new DynamicContentBlocks(editor, opts);
-    dcb.addDynamciContentBlock();
-  } // Add icon to mj-hero
+    dcb.addDynamciContentBlock();    
+  } else if ('grapesjswebpage' in cfg.pluginsOpts) {
```
Comment (adiux) [inline dist/blocks.js:14]: I think this can be solved with the function: ContentService.getMode(editor)

File: src/blocks.js [hunk 2]
```diff
@@ -11,8 +14,69 @@ export default (editor, opts = {}) => {
   } else if ('grapesjsnewsletter' in cfg.pluginsOpts) {
     const dcb = new DynamicContentBlocks(editor, opts);
     dcb.addDynamciContentBlock();
+  }else if ('grapesjswebpage' in cfg.pluginsOpts) {
```
Comment (adiux) [inline src/blocks.js]: I think this can be solved more nicely now with the new function: ContentService.getMode(editor) 

(the same is true for the existing code on one 14.)

File: src/blocks.js [hunk 3]
```diff
@@ -11,8 +14,69 @@ export default (editor, opts = {}) => {
...
+    pcb.addPreferenceCenterBlock();
+
+    bm.add('button', {
+      label: 'Button',
+      category : Mautic.translate('grapesjsbuilder.categoryBlockLabel'),
+      attributes: {
+        class: 'gjs-fonts gjs-f-button'
+      },
+      content: 
+        `<style>
```
Comment (adiux) [inline src/blocks.js]: The prefered way is to use a scss file for all styles

**Comment by disha1611** [comment]
@RCheesley @adiux some progress on pref ceneter block there is all block covered except where from dateand to date input requires. There is still some work needs to done like label for each block with grapejsbuilder Mautic generalise lable in all language also right now icons of all block is same. Need to have proper icons based on name.

**Comment by disha1611** [comment]
@adiux  if there is any code modification needed plz let me know. @RCheesley if you can check this blocks which are done. It will be really helpful. Guys if there some css related suggestion plz let me know. Thanks

**Comment by disha1611** [comment]
Hello @adiux @RCheesley I have added date type input for channel frequency block in preference center category. There is one issue when we drag Channel Frequency block for the first time into the canvas it will work fine for both from date and to date inputs. But if you drag it again the there is just simple input box not with the date type and date picker doesn't open at that place.

**Comment by RCheesley** [comment]
Thanks @disha-pishavadia24 for all the work on this!  It looks like you have some conflicts to address with this PR.

I don't believe anyone would drag the channel frequency twice, because you would only have that show once on the page. Or do you mean that they drag it onto the canvas, delete it, and then drag it on again?

**Comment by disha1611** [comment]
Yes like drag once delete and then again drag. and there will be a issue.

File: src/preferenceCenter/preferenceCenter.service.js [hunk 4]
```diff
@@ -0,0 +1,226 @@
+import Logger from 'grapesjs-preset-mautic/src/logger';
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.service.js]: Here is some code from the dynamic content. I don't think that should be here?

**Review by adiux — CHANGES_REQUESTED** [review]

File: src/preferenceCenter/preferenceCenter.service.js (hunk 4 above)
Comment (RCheesley) [reply src/preferenceCenter/preferenceCenter.service.js]: @disha-pishavadia24 can you take a look at the review please?

Comment (disha1611) [reply src/preferenceCenter/preferenceCenter.service.js]: @adiux @RCheesley Extra file removed. And that was not in consideration.

File: src/blocks.js [hunk 5]
```diff
@@ -32,6 +48,48 @@ export default (editor, opts = {}) => {
     });
   });
 
+  /*
+   * Blocks for Preference Center Category
+   */
+  //check if page is for preference center
+  if (mQuery('#page_isPreferenceCenter_1').is(':checked')) {
```
Comment (adiux) [inline src/blocks.js:57]: to make the code better readable and more robust this could be converted to a function. E.g.: isPreferenceCenter()

File: src/commands.js [hunk 6]
```diff
@@ -25,4 +26,29 @@ export default (editor) => {
...
+      // preferenceCenterCmd.convertPreferenceCenterTokensToSlots();
+      preferenceCenterCmd.launchPreferenceCenterPopup(edtr, sender, options);
+    },
+    stop: (edtr) => preferenceCenterCmd.stopPreferenceCenterPopup(edtr),
+  });
+  // Slot to {token}
+  editor.Commands.add('preset-mautic:preference-center-slots-to-tokens', {
+    run: (edtr) => preferenceCenterCmd.convertPreferenceCenterSlotsToTokens(edtr),
+  });
+  // {token} to slot
```
Comment (adiux) [inline src/commands.js]: for consistency with the grapesjs language we should rename Slots to Components

File: src/preferenceCenter/preferenceCenter.blocks.js [hunk 7]
```diff
@@ -0,0 +1,176 @@
+export default class PreferenceCenterBlocks {
+  blockManager;
+
+  constructor(editor) {
+    this.blockManager = editor.BlockManager;
+  }
+
+  addPreferenceCenterBlock() {
+    // add preference center category-list block
+    this.blockManager.add('MyCategories', {
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.blocks.js:10]: I guess this should be renamed to something more specific. E.g. Preference Center?

File: src/preferenceCenter/preferenceCenter.blocks.js [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+      tagName: 'div',
+      label: 'My Categories',
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.blocks.js:12]: see last comment

File: src/preferenceCenter/preferenceCenter.blocks.js [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+      category: Mautic.translate('grapesjsbuilder.preferenceCenterLabel'),
+      draggable: false,
+      media:
+        '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M22 5.5c0-.3-.5-.5-1.3-.5H3.4c-.8 0-1.3.2-1.3.5v3c0 .3.5.5 1.3.5h17.4c.8 0 1.3-.2 1.3-.5v-3zM21 8H3V6h18v2zM22 10.5c0-.3-.5-.5-1.3-.5H3.4c-.8 0-1.3.2-1.3.5v3c0 .3.5.5 1.3.5h17.4c.8 0 1.3-.2 1.3-.5v-3zM21 13H3v-2h18v2z"/><rect width="10" height="3" x="2" y="15" rx=".5"/></svg>',
+      content: {
+        components: [
+          {
+            components: [{ type: 'label', components: 'My Categories' }],
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.blocks.js:20]: same here

File: src/preferenceCenter/preferenceCenter.blocks.js [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+              { type: 'label', components: 'Category' },
+            ],
+          },
+        ],
+      },
+      style: { padding: '50px' },
+    });
+
+    //add my segment list
+    this.blockManager.add('MySegment', {
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.blocks.js:37]: All the labels need to be accurate and translatable

File: src/preferenceCenter/preferenceCenter.commands.js [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  /**
+   * Load Dynamic Content editor and append to the codePopup Modal
+   */
+  addPreferenceCenterEditor(editor, options) {
+    const { target } = options;
+    const dcComponent = target || editor.getSelected();
+    if (!dcComponent) {
+      throw new Error('No dc components found');
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.commands.js]: should be changed to "No preference center components found"

File: src/preferenceCenter/preferenceCenter.commands.js [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+      );
+    }
+
+    // Show if hidden
+    focusForm.removeClass('fade');
+    // Hide delete default button
+    focusForm.find('.tab-pane:first').find('.remove-item').hide();
+
+    // Insert inside popup
+    mQuery(this.dcPopup).empty().append(focusForm.detach());
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.commands.js]: "dc" stands for dynamic content. please change all occurrences to "prefereceCenter" or "pc"

Comment (adiux) [reply src/preferenceCenter/preferenceCenter.commands.js]: same for "dynCon"

File: src/preferenceCenter/preferenceCenter.domcomponents.js [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+
+              if (
+                typeof daterangepicker == 'undefined' ||
+                typeof jQuery == 'undefined' ||
+                typeof moment == 'undefined'
+              ) {
+                if (typeof jQuery == 'undefined') {
+                  var jquery = document.createElement('script');
+                  jquery.src =
+                    '//cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js';
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.domcomponents.js]: I think jquery is already loaded with Mautic. Do we really need to load it again?

**Review by adiux — CHANGES_REQUESTED** [review]
Looking good. Lets bring this home! Just a couple of things in the comments

File: src/blocks.js (hunk 5 above)
Comment (disha1611) [reply src/blocks.js:57]: Hi @adiux , its done

File: src/preferenceCenter/preferenceCenter.blocks.js (hunk 10 above)
Comment (disha1611) [reply src/preferenceCenter/preferenceCenter.blocks.js:37]: Okay yes that point is in my mind. I will do this.

File: src/preferenceCenter/preferenceCenter.blocks.js (hunk 9 above)
Comment (disha1611) [reply src/preferenceCenter/preferenceCenter.blocks.js:20]: For My Category its the same as it was in legacy builder. I gave the same name for all labels from legacy builder.

File: src/preferenceCenter/preferenceCenter.blocks.js (hunk 7 above)
Comment (disha1611) [reply src/preferenceCenter/preferenceCenter.blocks.js:10]: I can change block's name while we add using blockManager.add() but the label of that is "My Category" and I gave same name from the legacy builder. Preference Ceneter is category. Like we have "section" , "block" for preference center pages we will have "Prefernce Center" and when you open it under that you will find Preference ceneter related blocks MyCategory, My Segement, Save Preference etc. If we want to change, please let me know what that label would be.

File: src/preferenceCenter/preferenceCenter.commands.js (hunk 11 above)
Comment (disha1611) [reply src/preferenceCenter/preferenceCenter.commands.js]: this file is removed as there is no use as of now

File: src/preferenceCenter/preferenceCenter.domcomponents.js (hunk 13 above)
Comment (disha1611) [reply src/preferenceCenter/preferenceCenter.domcomponents.js]: It's a component specific requirement as to load datepicker into the canvas we will need it. I think the jquery loaded by mautic is general for all pages of the project. But inside the builder if its already loaded then it will not give undefined and rest of the code will work.

Comment (disha1611) [reply src/preferenceCenter/preferenceCenter.domcomponents.js]: @adiux the issue with this code is when we drag date element first time it's get loaded properly date input works fine. But if we delete the element and drag it again to canvas then it does not open datepicker. If you can look into the error that will be really helpful.  The name of the block which contains the date component is Channel Frequency. Thanks

**Comment by RCheesley** [comment]
@disha-pishavadia24 looks like some conflicts to resolve and suggestions/comments in code review. Nearly there! :)

**Comment by disha1611** [comment]
Will update soon @RCheesley

File: src/blocks.js (hunk 5 above)
Comment (adiux) [reply src/blocks.js:57]: there is another occurence of this check on line 56

File: src/blocks.js [hunk 14]
```diff
@@ -32,6 +49,48 @@ export default (editor, opts = {}) => {
...
+    if (typeof bm.get('MyCategories') !== 'undefined') {
```
Comment (adiux) [inline src/blocks.js:58]: This code is only used with the preferenceCenter. I think it makes sense to move it to the place where the rest of the preference center code is: preferenceCenter/

File: src/preferenceCenter/preferenceCenter.commands.js [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+
+  // eslint-disable-next-line class-methods-use-this
+  stopPreferenceCenterPopup(editor) {
+    editor.runCommand('preset-mautic:dynamic-content-tokens-to-slots');
+  }
+
+  /**
+   * Convert all dynamic content tokens on the canvas to
+   * human readable texts/slots/component.
+   * {dynamiccontent} => Dynamic Content
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.commands.js]: please rename

File: src/preferenceCenter/preferenceCenter.commands.js [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+  // eslint-disable-next-line class-methods-use-this
+  convertPreferenceCenterSlotsToTokens(editor) {
+    // get all dynamic content elements loaded in the editor
+    const preferenceCenter = editor.DomComponents.getWrapper().find('[data-slot="preferenceCenter"]');
+    preferenceCenter.forEach((preferenceCenter) => {
+      const attributes = preferenceCenter.getAttributes();
+      const decId = attributes['data-param-dec-id'];
+      // If it's not a token -> convert to token
+      if (decId >= 0) {
+        const dynConId = PreferenceCenterCommands.getPcStoreId(attributes['data-param-dec-id']);
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.commands.js]: please rename all the old code like `dynCon` to something that identifies the preference center

File: src/preferenceCenter/preferenceCenter.commands.js [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+
+        const dynConTarget = mQuery(dynConId);
+        const dynConName = dynConTarget.find(`${dynConId}_tokenName`).val();
+        const dynConToken = `{dynamiccontent="${dynConName}"}`;
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.commands.js]: can you help me out here? What is it doing with the dynamic content token?

File: src/preferenceCenter/preferenceCenter.commands.js [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+
+    this.addPreferenceCenterEditor(editor, options);
+
+    const title = Mautic.translate('grapesjsbuilder.preferenceCenterBlockLabel');
+    const modal = editor.Modal;
+    modal.setTitle(title);
+
+    editor.Modal.setContent(this.dcPopup);
+    modal.open();
+    modal.onceClose(() => editor.stopCommand('preset-mautic:dynamic-content-open'));
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.commands.js]: I guess this should stop the preference center modal, not the dynamic content modal. This could be the fix for the problem that you can not open the modal multiple times.

File: src/preferenceCenter/preferenceCenter.commands.js [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  /**
+   * Build the basic popup/modal frame to hold the dynamic content editor
+   * @returns HTMLDivElement
+   */
+  static buildPreferenceCenterPopup() {
+    const content = document.createElement('div');
+    content.setAttribute('id', 'dynamic-content-popup');
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.commands.js]: again: dynamic content instead of preference center

File: src/preferenceCenter/preferenceCenter.commands.js (hunk 11 above)
Comment (adiux) [reply src/preferenceCenter/preferenceCenter.commands.js]: please remove the file then

**Review by adiux — CHANGES_REQUESTED** [review]
Please create a final PR with all the changes to be reviewed.

File: src/preferenceCenter/preferenceCenter.service.js (hunk 4 above)
Comment (adiux) [reply src/preferenceCenter/preferenceCenter.service.js]: the file preferenceCenter.service.js still talks about `dynamic content`. Maybe it is just the comments and the variables that are named wrong? But this is very confusing. Please adjust it accordingly.

**Review by adiux — CHANGES_REQUESTED** [review]

File: src/preferenceCenter/preferenceCenter.commands.js (hunk 15 above)
Comment (adiux) [reply src/preferenceCenter/preferenceCenter.commands.js]: still dynamic-content used. Please adjust

**Comment by disha1611** [comment]
Will check for sure! Thanks Adrian.

**Comment by disha1611** [comment]
@adiux @RCheesley  file removed conatining the dynamic center references.

**Comment by adiux** [comment]
Hi @disha1611 , what is your status on this? Can you please resolve the conflicts? And I guess it will also need a rebase to reflect the latest changes. on the main branch.

File: src/preferenceCenter/preferenceCenter.blocks.js [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+      content: {
+        // type: 'PreferredChannel',
+        components: [
+          {
+            components: [
+              {
+                type: 'checkbox',
+                attributes: { value: 1, checked: 'checked' },
+              },
+              { type: 'label', components: 'I want to receive %channel%' },
```
Comment (adiux) [inline src/preferenceCenter/preferenceCenter.blocks.js:122]: all these labels need to use the translations

**Comment by kuzmany** [comment]
We can close it favor to https://github.com/mautic/grapesjs-preset-mautic/pull/8

**Comment by RCheesley** [comment]
Closing PR in favour of #29
